File: client/ios/ProstoVPN/ServerSheet.swift
import SwiftUI

struct CurrentServerCard: View {
    @EnvironmentObject private var state: AppState
    let onOpen: () -> Void

    var body: some View {
        Button(action: onOpen) {
            ServerRow(server: state.currentServer) {
                Image(systemName: "chevron.up")
                    .font(.system(size: 15, weight: .semibold))
                    .foregroundColor(Theme.textTertiary)
            }
            .padding(.vertical, 13)
            .padding(.horizontal, 16)
            .contentShape(RoundedRectangle(cornerRadius: 26, style: .continuous))
        }
        .glassCardButton(cornerRadius: 26)
    }
}

struct ServerListSheet: View {
    @EnvironmentObject private var state: AppState
    @Environment(\.dismiss) private var dismiss

    private var t: L10n { state.t }

    var body: some View {
        ScrollView(showsIndicators: false) {
            VStack(alignment: .leading, spacing: 2) {
                ForEach(Array(state.displayServers().enumerated()), id: \.offset) { index, server in
                    row(server: server, index: index)
                }
            }
            .padding(.horizontal, 20)
            .padding(.bottom, 12)
        }
        .softScrollEdge()
        .safeAreaInset(edge: .top, spacing: 0) {
            Text(t.chooseServer)
                .font(.manrope(13, .bold))
                .kerning(0.5)
                .foregroundColor(Theme.textMuted)
                .frame(maxWidth: .infinity, alignment: .leading)
                .padding(.horizontal, 28)
                .padding(.top, 22)
                .padding(.bottom, 8)
        }
    }

    private func row(server: DisplayServer, index: Int) -> some View {
        let isActive = index == state.selectedServerIndex
        return Button {
            Haptics.selection()
            state.selectedServerIndex = index
            dismiss()
        } label: {
            ServerRow(server: server) {
                if isActive {
                    Image(systemName: "checkmark")
                        .font(.system(size: 16, weight: .bold))
                        .foregroundColor(Theme.link)
                }
            }
            .padding(.vertical, 12)
            .padding(.horizontal, 8)
            .background(isActive ? Theme.accent.opacity(0.07) : .clear)
            .clipShape(RoundedRectangle(cornerRadius: 16, style: .continuous))
            .contentShape(Rectangle())
        }
        .buttonStyle(ScaleButtonStyle(scale: 0.98))
    }
}

struct ServerRow<Trailing: View>: View {
    let server: DisplayServer?
    @ViewBuilder var trailing: () -> Trailing

    var body: some View {
        HStack(spacing: 14) {
            FlagChip(flag: server?.flag ?? "🌐")

            VStack(alignment: .leading, spacing: 1) {
                HStack(spacing: 8) {
                    Text(server?.name ?? "—")
                        .font(.manrope(15, .bold))
                        .foregroundColor(Theme.text)
                        .lineLimit(1)

                    ProtocolBadge()
                }

                Text(server?.sub ?? "")
                    .font(.manrope(12.5, .medium))
                    .foregroundColor(Theme.textMuted)
                    .lineLimit(1)
            }

            Spacer()

            trailing()
        }
    }
}

File: client/windows/src/main/kotlin/com/prostovpn/desktop/ServerSheet.kt
package com.prostovpn.desktop

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

@Composable
fun CurrentServerCard(
    state: AppState,
    backdrop: BackdropState,
    onOpen: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val interaction = remember { MutableInteractionSource() }
    Box(
        modifier = modifier
            .fillMaxWidth()
            .pressScale(interaction, 0.98f)
            .clip(RoundedCornerShape(26.dp))
            .liquidGlass(backdrop, cornerRadius = 26.dp)
            .pressHighlight(interaction, 0.05f, cornerRadius = 26.dp)
            .clickable(interactionSource = interaction, indication = null, onClick = onOpen)
            .padding(vertical = 13.dp, horizontal = 16.dp),
    ) {
        ServerRow(server = state.currentServer) {
            Icon(
                imageVector = Icons.chevronUp,
                contentDescription = null,
                tint = Theme.textTertiary,
                modifier = Modifier.size(17.dp),
            )
        }
    }
}

@Composable
fun ServerListSheet(
    state: AppState,
    visible: Boolean,
    backdrop: BackdropState,
    onDismiss: () -> Unit,
) {
    val haptics = rememberHaptics()
    val s = state.s

    GlassSheet(visible = visible, backdrop = backdrop, onDismiss = onDismiss) {
        Text(
            text = s.chooseServer,
            style = manrope(12.5.sp, W.bold, Theme.textMuted, letterSpacing = 0.6.sp),
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = Layout.screenPadding)
                .padding(top = 14.dp, bottom = 10.dp),
        )

        Column(
            modifier = Modifier
                .fillMaxWidth()
                .heightIn(max = 320.dp)
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 14.dp)
                .padding(bottom = Layout.screenPadding),
            verticalArrangement = Arrangement.spacedBy(4.dp),
        ) {
            state.displayServers().forEachIndexed { index, server ->
                val isActive = index == state.selectedServerIndex
                Box(
                    modifier = Modifier
                        .fillMaxWidth()
                        .scaleClickable(0.98f) {
                            haptics.selection()

                            state.switchServer(index)
                            onDismiss()
                        }
                        .clip(RoundedCornerShape(18.dp))
                        .background(
                            if (isActive) Theme.accent.copy(alpha = 0.10f) else Color.Transparent
                        )
                        .padding(vertical = 11.dp, horizontal = 10.dp),
                ) {
                    ServerRow(server = server) {
                        if (isActive) {
                            Icon(
                                imageVector = Icons.check,
                                contentDescription = null,
                                tint = Theme.link,
                                modifier = Modifier.size(18.dp),
                            )
                        }
                    }
                }
            }
        }
    }
}

@Composable
fun ServerRow(
    server: DisplayServer?,
    trailing: @Composable () -> Unit,
) {
    Row(
        modifier = Modifier.fillMaxWidth(),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        FlagChip(flag = server?.flag ?: "🌐")

        Spacer(Modifier.width(14.dp))

        Column(verticalArrangement = Arrangement.spacedBy(1.dp)) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Text(
                    text = server?.name ?: "—",
                    style = manrope(15.sp, W.bold, Theme.text),
                    maxLines = 1,
                )
                Spacer(Modifier.width(8.dp))
                ProtocolBadge()
            }

            Text(
                text = server?.sub ?: "",
                style = manrope(12.5.sp, W.medium, Theme.textMuted),
                maxLines = 1,
            )
        }

        Spacer(Modifier.weight(1f))

        trailing()
    }
}
